emf/src/main/kotlin/com/strumenta/kolasu/emf/cli/EMFCLICommands.kt

```diff
@@ -0,0 +1,106 @@
+        name = "emfmodel"
+    ) where P : EMFMetamodelSupport, P : ASTParser<R> {
+    val metamodel by option("--metamodel")
+    val outputDirectory by option("--output", "-o")
+        .file()
+        .help("Directory where to store the output. By default the current directory is used")
+        .default(File("."))
+    // val addMetamodel by option("--add-metamodel", "-amm").flag(default = false)
+    val includeKolasu by option("--include-kolasu", "-ik").flag(default = false)
+    val includeMetamodel by option("--include-metamodel", "-imm").flag(default = false)
+
+    override fun finalizeRun() {
+        // Nothing to do
+    }
+
+    override fun processException(input: File, relativePath: String, e: Exception) {
+        // Nothing to do
+    }
+
+    override fun processResult(input: File, relativePath: String, result: ParsingResult<R>, parser: P) {
+        val targetFile = File(this.outputDirectory.absolutePath + File.separator + relativePath)
+            .changeExtension("json")
+        val targetFileParent = targetFile.parentFile
+        targetFileParent.absoluteFile.mkdirs()
+
+        val resource = saveMetamodel(parser, targetFile)
+        val start = System.currentTimeMillis()
+        if (verbose) {
+            echo("Saving AST for $input to $targetFile... ")
+        }
+        result!!.saveModel(resource, URI.createFileURI(targetFile.path), includeMetamodel = includeMetamodel)
+        if (verbose) {
+            echo("Done (${System.currentTimeMillis() - start}ms).")
+        }
+    }
+
+    private fun saveMetamodel(parser: EMFMetamodelSupport, target: File): Resource {
+        val start = System.currentTimeMillis()
+        val metamodel = if (this.metamodel != null) {
+            File(this.metamodel!!)
+        } else {
+            File(target.parentFile, "metamodel." + (target.path.substring(target.path.lastIndexOf(".") + 1)))
+        }.absolutePath
+        if (verbose) {
+            echo("Saving metamodel to $metamodel... ")
+        }
+        val mmResource = parser.saveMetamodel(URI.createFileURI(target.path))
+        if (verbose) {
+            echo("Done (${System.currentTimeMillis() - start}ms).")
+        }
+        return mmResource
+    }
+}
+
+class EMFMetaModelCommand(val metamodelGenerator: Supplier<EPackage>) :
+    CliktCommand(
+        help = "Generate the metamodel for a language.",
+        name = "emfmetamodel"
+    ) {
+    val verbose by option("--verbose", "-v")
+        .help("Print additional messages")
+        .flag(default = false)
+    val output by option("--output", "-o")
+        .file()
+        .help("File where to store the metamodel")
+        .default(File("metamodel.json"))
+    val includeKolasu by option("--include-kolasu", "-ik").flag(default = false)
+
+    override fun run() {
+        val metamodel = metamodelGenerator.get()
+        val mmResource = JsonResourceFactory().createResource(URI.createFileURI(output.path))
```

Comment:
The original ParserCLIEMF used the resource returned from `EMFMetamodelSupport.saveMetamodel` whose `ResourceFactory` is initialized like this:
```
resourceSet.resourceFactoryRegistry.extensionToFactoryMap["json"] = JsonResourceFactory()
    resourceSet.resourceFactoryRegistry.extensionToFactoryMap["xmi"] = XMIResourceFactoryImpl()
    resourceSet.resourceFactoryRegistry.extensionToFactoryMap["ecore"] = EcoreResourceFactoryImpl()
```

So it can save to .json, .xmi and .ecore. I think we should preserve at least .json and .xmi, not sure about .ecore (maybe @loradd can comment on the difference between the two? I seem to recall that .ecore only supports one EPackage and silently drops the others).

Reply:
Good point. I created utility methods for creating resourceSets and resources and I tried to use them consistently in the project

Reply:
> (maybe @loradd can comment on the difference between the two? I seem to recall that .ecore only supports one EPackage and silently drops the others).

Not sure about this, I always used `XMIResourceFactory`.

Looking at the code, both factories seem to create resources extending `XMLResource`.  `EcoreResourceFactory` seems to configure details about IDs, URIs and encoding, while `XMIResource` is an extension of `XMLResource` in `XMIResourceFactory`.

```java
// EcoreResourceFactoryImpl
@Override
public Resource createResource(URI uri) {
    XMLResource result = new XMLResourceImpl(uri) {
        @Override
         protected boolean useIDs() {
             return eObjectToIDMap != null || idToEObjectMap != null;
         }
     };
     result.setEncoding("UTF-8");
     result.getDefaultSaveOptions().put(XMLResource.OPTION_USE_ENCODED_ATTRIBUTE_STYLE, Boolean.TRUE);
     result.getDefaultSaveOptions().put(XMLResource.OPTION_LINE_WIDTH, 80);
     result.getDefaultSaveOptions().put(XMLResource.OPTION_URI_HANDLER, new URIHandler.PlatformSchemeAware());
        return result;
}

// XMIResourceFactoryImpl
@Override
public Resource createResource(URI uri) {
    return new XMIResourceImpl(uri);
}
```